presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1354,11 +1356,26 @@ else if (expression instanceof LongLiteral) {
                     }
 
                     // otherwise, just use the expression as is
+                    boolean orderByOnExpression = false;
                     if (orderByExpression == null) {
                         orderByExpression = expression;
+                        orderByOnExpression = true;
                     }
 
                     ExpressionAnalysis expressionAnalysis = analyzeExpression(orderByExpression, sourceScope);
+                    if (orderByOnExpression) {
+                        // Try to resolve within orderByScope. If it is possible, then we warn - standard behavior differs from legacy one.
+                        for (Map.Entry<Expression, FieldId> columnExpression : expressionAnalysis.getColumnReferences().entrySet()) {
+                            Optional<ResolvedField> resolvedField = orderByScope.tryResolveField(columnExpression.getKey());
+                            if (resolvedField.isPresent()) {
+                                if (!FieldId.from(resolvedField.get()).equals(columnExpression.getValue())) {
```

Comment:
OK. Now I see. It seems that `expressionAnalysis` doesn't contain information about subquries. Therefore we won't provide user with warning for query like:
```
SELECT 1 as a FROM (VALUES 1) t(a) ORDER BY (SELECT a FROM (VALUES 2) t(b));
```
Please add tests:
```
(no warning, unsupported) SELECT 1 as a FROM (VALUES 1) t(a) ORDER BY (SELECT a FROM (VALUES 2) t(b))
(warning) SELECT 'a' as a FROM (VALUES 1) t(a) ORDER BY transform(array[1], x->a)[1];
(no warning) SELECT 'a' as a FROM (VALUES 1) t(a) ORDER BY transform(array[1], a->a)[1];
```